
# Exploring eBay Car Sales Data

In this project I'll be cleaning, preparing, and researching used car sales data. The data is sourced from a German eBay site (*eBay Kleinanzeigan*).

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

Looks like we have a count of 50,000 rows and 20 columns to work with. I'm going to definitely make sure the `dateCrawled` index is in a consistent format. `name` will need to be broken down into make, model, and manufacturer (which looks synonymous with `brand`).  

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

Columns `price` and `odometer` will need to be converted to `int64` or `float64` as needed. It will also be nice to change all our column headers to a more consistently pythonic version. There is also quite a bit of missing info in certain columns (particularly `notRepairedDamage`).

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

**Did Pandas add an index column or do I need to do that?**

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Columns are `camelCase` and not our lovely, Pythonic `snakecase`. We'll also convert oddly named columns like `nrOfPictures` to `picture_count` and the ambiguous `dateCreated` to `ad_created`.

In [5]:
mapping_col_dict = {'dateCrawled': 'date_crawled',
                    'offerType': 'offer_type',
                    'vehicleType': 'vehicle_type',
                    'yearOfRegistration': 'registration_year',
                    'powerPS': 'power_ps',
                    'monthOfRegistration': 'registration_month',
                    'fuelType': 'fuel_type',
                    'notRepairedDamage': 'unrepaired_damage',
                    'dateCreated': 'ad_created',
                    'nrOfPictures': 'picture_count',
                    'postalCode': 'postal_code',
                    'lastSeen': 'last_seen'
                   }

autos = autos.rename(mapping_col_dict, axis="columns")

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  picture_count  \
0     peugeot              nein  2016-03-26 00:00:00              0   
1         bmw              nein  2016-04-04 00:00:00              0   
2  volkswagen              nein  2016-03-26 00:00:00              0   
3       smart              nein  2016-03-12 00:00:00              0   
4        ford              nein  2016-04-01 00:00:00              0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Much better! :)

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  picture_count   postal_code            last_seen  
count                 50000        50000.0  50000.000000                50000  
unique                   76            NaN           NaN                39481  
top     2016-04-03 00:00:00            NaN           NaN  2016-04-07 06:17:27  
freq                   1946            NaN           NaN                    8  
mean                    NaN            0.0  50813.627300                  NaN  
std                     NaN            0.0  25779.747957                  NaN  
min                     NaN            0.0   1067.000000                  NaN  
25%                     NaN            0.0  30451.000000                  NaN  
50%                     NaN            0.0  49577.000000                  NaN  
75%                     NaN            0.0  71540.000000                  NaN  
max                     NaN            0.0  99998.000000                  NaN

In [7]:
autos[~(autos['seller'] == 'privat')]

date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type price   abtest vehicle_type  registration_year  \
7738  gewerblich    Angebot  $100  control        kombi               2000   

      gearbox  power_ps   model   odometer  registration_month fuel_type  \
7738  manuell         0  megane  150,000km                   8    benzin   

        brand unrepaired_damage           ad_created  picture_count  \
7738  renault               NaN  2016-03-15 00:00:00              0   

      postal_code            last_seen  
7738        65232  2016-04-06 17:15:37

In [8]:
autos[~(autos['offer_type'] == 'Angebot')]

date_crawled                  name  seller offer_type price  \
17541  2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch    $0   

      abtest vehicle_type  registration_year gearbox  power_ps        model  \
17541   test          bus               2005     NaN         0  transporter   

        odometer  registration_month fuel_type       brand unrepaired_damage  \
17541  150,000km                   0       NaN  volkswagen               NaN   

                ad_created  picture_count  postal_code            last_seen  
17541  2016-04-03 00:00:00              0        29690  2016-04-05 15:16:06

In [9]:
autos[~(autos['unrepaired_damage'] == 'nein')]

date_crawled                                               name  \
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
...                    ...                                                ...   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49989  2016-03-11 19:50:37        VW_Polo_zum_Ausschlachten_oder_Wiederaufbau   
49991  2016-03-06 15:25:19                                         Kleinwagen   
49993  2016-03-15 18:47:35           Audi_A3__1_8l__Silber;_schoenes_Fahrzeug   

       seller offer_type   price   abtest vehicle_type  registration_year  \
5      privat    Angebot  $7,900     test          bus               2006   
6      privat    Angebot    $300     test    limousine               1995   
11     privat    Angebot    $350  control          NaN               2000   
14     privat    Angebot  $3,999     test   kleinwagen               2007   
18     privat    Angebot    $300  control          bus               1995   
...       ...        ...     ...      ...          ...                ...   
49984  privat    Angebot      $0     test          NaN               2000   
49985  privat    Angebot  $1,000  control          NaN               1995   
49989  privat    Angebot    $150     test   kleinwagen               1997   
49991  privat    Angebot    $500  control          NaN               2016   
49993  privat    Angebot  $1,650  control   kleinwagen               1997   

         gearbox  power_ps    model   odometer  registration_month fuel_type  \
5      automatik       150  voyager  150,000km                   4    diesel   
6        manuell        90     golf  150,000km                   8    benzin   
11           NaN         0      NaN  150,000km                   0    benzin   
14       manuell        75     clio  150,000km                   9    benzin   
18       manuell       150      3er  150,000km                   0    benzin   
...          ...       ...      ...        ...                 ...       ...   
49984        NaN         0      NaN  150,000km                   0       NaN   
49985  automatik         0      NaN  150,000km                   0    benzin   
49989    manuell         0     polo  150,000km                   5    benzin   
49991    manuell         0   twingo  150,000km                   0    benzin   
49993    manuell         0      NaN  150,000km                   7    benzin   

                brand unrepaired_damage           ad_created  picture_count  \
5            chrysler               NaN  2016-03-21 00:00:00              0   
6          volkswagen               NaN  2016-03-20 00:00:00              0   
11      mercedes_benz               NaN  2016-03-16 00:00:00              0   
14            renault               NaN  2016-03-23 00:00:00              0   
18                bmw               NaN  2016-03-26 00:00:00              0   
...               ...               ...                  ...            ...   
49984  sonstige_autos               NaN  2016-03-31 00:00:00              0   
49985      volkswagen               NaN  2016-04-02 00:00:00              0   
49989      volkswagen                ja  2016-03-11 00:00:00              0   
49991         renault               NaN  2016-03-06 00:00:00              0   
49993            audi               NaN  2016-03-15 00:00:00              0   

       postal_code            last_seen  
5            22962  2016-04-06 09:45:21  
6            31535  2016-03-23 02:48:59  
11           17498  2016-03-16 18:45:34  
14           81737  2016-04-01 15:46:4

Taking a closer look at the data that needs cleaning.  
`seller` and `offer_type` only have two unique entries. One of them has 49,999 occurences. We can probably drop these columns altogether.  
`unrepaired_damage` has three different entry types. That will require more investigation.  
`registration_year`, `power_ps`, `postal_code`, and `picture_count` need to be converted to `int`.  
`price` and `odometer` are currently string objects. They need to be stripped of `$` and converted to `int`.
All the date specific columns have a date format. Gonna have to make sure those become the correct `dtype`.  
`registration_month` has values of 1000 and 9999. Those are certianly not years that are possible.

### Dropping Useless Columns

I'm going to start by dropping the useless columns containing 99.99% the same data.  
`seller` has `privat` for all but one entry so that can go.  
`offer_type` has 49,999 `Angebot`. That is not useful. Drop it.  
`picture_count` has 0's and NaN's. Seeya.

In [10]:
autos.drop(['seller', 'offer_type', 'picture_count'], axis=1, inplace=True)

autos.describe(include='all')

date_crawled         name  price abtest vehicle_type  \
count                 50000        50000  50000  50000        44905   
unique                48213        38754   2357      2            8   
top     2016-04-02 11:37:04  Ford_Fiesta     $0   test    limousine   
freq                      3           78   1421  25756        12859   
mean                    NaN          NaN    NaN    NaN          NaN   
std                     NaN          NaN    NaN    NaN          NaN   
min                     NaN          NaN    NaN    NaN          NaN   
25%                     NaN          NaN    NaN    NaN          NaN   
50%                     NaN          NaN    NaN    NaN          NaN   
75%                     NaN          NaN    NaN    NaN          NaN   
max                     NaN          NaN    NaN    NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN


### `price` and `odometer`


**Clean `price` column, convert to `int64`.**  
`dtype` for `value_count` being `int64` refers to the `value_count` as it's OWN object.

In [11]:
print(autos["price"].value_counts())
print(autos["price"].describe())

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$414          1
$79,933       1
$5,198        1
$18,890       1
$16,995       1
Name: price, Length: 2357, dtype: int64
count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object


In [12]:
autos["price"] = autos["price"].str.replace('$', '', regex=False).str.replace(',', '', regex=False).astype(int)
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
414         1
79933       1
5198        1
18890       1
16995       1
Name: price, Length: 2357, dtype: int64

**Clean `odometer` column, convert to `int64`.**  
Rename column as `odometer_km`.

In [13]:
print(autos["odometer"].value_counts())
print(autos["odometer"].describe())

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64
count         50000
unique           13
top       150,000km
freq          32424
Name: odometer, dtype: object


In [14]:
autos["odometer"] = autos["odometer"].str.replace('km', '', regex=False).str.replace(',', '', regex=False).astype(int)
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [15]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns.

In [16]:
prices = autos["price"]

# print(prices.unique().shape, '\n')
# print(prices.describe(), '\n')
# print(prices.value_counts().head().sort_index(ascending=False), '\n')
# print(prices.value_counts().head(25).sort_index(ascending=True), '\n')
# print(prices.value_counts().sort_index(ascending=False).head(), '\n')

prices.value_counts(bins=[0, 1, 100, 1000, 5000, 10000, 50000, 150000, 350000, 99999999]).sort_index()

(-0.001, 1.0]              1577
(1.0, 100.0]                319
(100.0, 1000.0]           10103
(1000.0, 5000.0]          20993
(5000.0, 10000.0]          8935
(10000.0, 50000.0]         7873
(50000.0, 150000.0]         166
(150000.0, 350000.0]         20
(350000.0, 99999999.0]       14
Name: price, dtype: int64

A starting bid of 1 dollar is not unheard of on auction sites, but I find it hard to believe that anyone is a listing a 350,000 dollar car on an eBay site. We'll remove everything over 351,000.

In [17]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [18]:
odometers = autos["odometer_km"]

# print(odometers.unique().shape, '\n')
# print(odometers.describe(), '\n')
print(odometers.value_counts().sort_index(ascending=True).head(25), '\n')
print(odometers.value_counts(bins=[0, 4999, 50000, 100000, 150000, 300000]).sort_index(), '\n')



5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64 

(-0.001, 4999.0]            0
(4999.0, 50000.0]        4458
(50000.0, 100000.0]      7636
(100000.0, 150000.0]    36471
(150000.0, 300000.0]        0
Name: odometer_km, dtype: int64 




### `last_seen`, `ad_created`, and `last_seen`


In [19]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [20]:
autos['date_crawled'].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object

In [21]:
date_crawled = autos['date_crawled'].str[:10]
date_crawled.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

It appears that this was all taken from a period of four weeks.

In [22]:
ad_created = autos['ad_created'].str[:10]
ad_created.value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

Ads were created as far back as 2015 June to our most recent 2016 April.

In [23]:
last_seen = autos['last_seen'].str[:10]
last_seen.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Something was happening on 2016-04-06 to cause a spike in last seen view count.


### `registration_year`


In [24]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The `registration_year` column has values ranging `1000` to `9999`. A car can't be registered after the date of the listing (2016). So these are incorrect.  
As for the other direction, it's fairly improbable for a car to have been registered before the first few decades of the 1900s. So we'll take a look at the rows outside `1900-2016` and see if we can just drop these rows.  

In [25]:
reg_year = autos["registration_year"]

pre_1900 = autos[reg_year < 1900]
# pre_1900.describe(include='all')

pre_1900

date_crawled                                   name  price  \
10556  2016-04-01 06:02:10                             UNFAL_Auto    450   
22316  2016-03-29 16:56:41  VW_Kaefer.__Zwei_zum_Preis_von_einem.   1500   
24511  2016-03-17 19:45:11            Trabant__wartburg__Ostalgie    490   
32585  2016-04-02 16:56:39                             UNFAL_Auto    450   
49283  2016-03-15 18:38:53                             Citroen_HY   7750   

        abtest vehicle_type  registration_year  gearbox  power_ps   model  \
10556  control          NaN               1800      NaN      1800     NaN   
22316  control          NaN               1000  manuell         0  kaefer   
24511  control          NaN               1111      NaN         0     NaN   
32585  control          NaN               1800      NaN      1800     NaN   
49283  control          NaN               1001      NaN         0  andere   

       odometer_km  registration_month fuel_type       brand  \
10556         5000                   2       NaN  mitsubishi   
22316         5000                   0    benzin  volkswagen   
24511         5000                   0       NaN     trabant   
32585         5000                   2       NaN  mitsubishi   
49283         5000                   0       NaN     citroen   

      unrepaired_damage           ad_created  postal_code            last_seen  
10556              nein  2016-04-01 00:00:00        63322  2016-04-01 09:42:30  
22316               NaN  2016-03-29 00:00:00        48324  2016-03-31 10:15:28  
24511               NaN  2016-03-17 00:00:00        16818  2016-04-07 07:17:29  
32585              nein  2016-04-02 00:00:00        63322  2016-04-04 14:46:21  
49283               NaN  2016-03-15 00:00:00        66706  2016-04-06 18:47:20

Only 6 rows before 1900. There are `NaN` values for `vehicle_type` and`gearbox`, and odometer readings that make no sense. Also these rows are claiming they were registered before the companies were even founded.

In [26]:
post_2016 = autos[reg_year > 2016]
post_2016.describe(include='all')

date_crawled        name         price abtest vehicle_type  \
count                  1879        1879   1879.000000   1879            2   
unique                 1876        1818           NaN      2            2   
top     2016-03-29 20:38:23  Ford_Focus           NaN   test        coupe   
freq                      2           4           NaN    957            1   
mean                    NaN         NaN   3693.304417    NaN          NaN   
std                     NaN         NaN   4873.190245    NaN          NaN   
min                     NaN         NaN      1.000000    NaN          NaN   
25%                     NaN         NaN   1000.000000    NaN          NaN   
50%                     NaN         NaN   2000.000000    NaN          NaN   
75%                     NaN         NaN   4500.000000    NaN          NaN   
max                     NaN         NaN  64900.000000    NaN          NaN   

        registration_year  gearbox      power_ps model    odometer_km  \
count         1879.000000     1650   1879.000000  1617    1879.000000   
unique                NaN        2           NaN   179            NaN   
top                   NaN  manuell           NaN  golf            NaN   
freq                  NaN     1386           NaN   193            NaN   
mean          2052.345929      NaN     98.307610   NaN  130643.959553   
std            444.992920      NaN    433.249342   NaN   37393.589236   
min           2017.000000      NaN      0.000000   NaN    5000.000000   
25%           2017.000000      NaN      0.000000   NaN  125000.000000   
50%           2017.000000      NaN     75.000000   NaN  150000.000000   
75%           2018.000000      NaN    120.000000   NaN  150000.000000   
max           9999.000000      NaN  16011.000000   NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count          1879.000000      1171        1879              1088   
unique                 NaN         5          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN       827         473               939   
mean              4.680681       NaN         NaN               NaN   
std               3.884680       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               1.000000       NaN         NaN               NaN   
50%               4.000000       NaN         NaN               NaN   
75%               8.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                  1879   1879.000000                 1879  
unique                   39           NaN                 1866  
top     2016-03-28 00:00:00           NaN  2016-04-07 08:45:02  
freq                     78           NaN                    2  
mean                    NaN  47950.638105                  NaN  
std                     NaN  25373.285156                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  27671.500000                  NaN  
50%                     NaN  45966.000000                  NaN  
75%                     NaN  66450.000000                  NaN  
max                     NaN  99974.000000                  NaN

In [27]:
# Remove the rows outside 1900-2016.
autos = autos[autos["registration_year"].between(1900, 2016)]

autos.describe()

price  registration_year      power_ps    odometer_km  \
count   46681.000000       46681.000000  46681.000000   46681.000000   
mean     5977.716801        2002.910756    117.892933  125586.855466   
std      9177.909479           7.185103    184.922911   39852.528628   
min         1.000000        1910.000000      0.000000    5000.000000   
25%      1250.000000        1999.000000     75.000000  100000.000000   
50%      3100.000000        2003.000000    109.000000  150000.000000   
75%      7500.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        2016.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        46681.000000  46681.000000  
mean             5.827125  51097.434181  
std              3.670300  25755.387192  
min              0.000000   1067.000000  
25%              3.000000  30827.000000  
50%              6.000000  49828.000000  
75%              9.000000  71732.000000  
max             12.000000  99998.000000

We removed 1972 rows from 50,000 leaving us 48028. Our ranges seem far more plausible now. With the vast majority beteween `1999-2008`.

In [28]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, Length: 78, dtype: float64


### `brand`


In [29]:
autos['brand'].describe()

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

In [30]:
brands_count = autos["brand"].value_counts(normalize=True)
round(brands_count * 100, 2)

volkswagen        21.13
bmw               11.00
opel              10.76
mercedes_benz      9.65
audi               8.66
ford               6.99
renault            4.71
peugeot            2.98
fiat               2.56
seat               1.83
skoda              1.64
nissan             1.53
mazda              1.52
smart              1.42
citroen            1.40
toyota             1.27
hyundai            1.00
sonstige_autos     0.98
volvo              0.91
mini               0.88
mitsubishi         0.82
honda              0.78
kia                0.71
alfa_romeo         0.66
porsche            0.61
suzuki             0.59
chevrolet          0.57
chrysler           0.35
dacia              0.26
daihatsu           0.25
jeep               0.23
subaru             0.21
land_rover         0.21
saab               0.16
jaguar             0.16
daewoo             0.15
trabant            0.14
rover              0.13
lancia             0.11
lada               0.06
Name: brand, dtype: float64

It looks like the top brand is volkswagen by a lot. The following big dogs are bmw, opel, mercedes_benz and audi. Let's see the mean prices for the brands above 5% frequency.

In [31]:
common_brands = brands_count[brands_count > 0.05].index

top_brand_avg = {}

for b in common_brands:
    b_rows = autos[autos['brand'] == b]
    b_prices = b_rows['price']
    mean_price = b_prices.mean()
    top_brand_avg[b] = int(mean_price)

pd.DataFrame.from_dict(top_brand_avg, orient='index').sort_values(by=0, ascending=False)

0
audi           9336
mercedes_benz  8628
bmw            8332
volkswagen     5402
ford           3749
opel           2975

After looking at the mean car prices for the most commonly listed brands it looks the higher end vehicles are Audi, Mercedes Benz, and BMW.  
The most affordable cars would be Ford and Opel.  
Volkswagen is like the best of both worlds right in the middle. Also the fact they account for 20% of all listings if probably wouldn't be too difficult to find for a very reasonable price.